<a href="https://colab.research.google.com/github/RMoulla/IYT/blob/main/TP_Seloger_Analyse_Donn%C3%A9es.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Travaux pratiques : analyse des prix des biens immobiliers.

Étude d'un jeu de données comportant des informations immobilières provenant du site seloger.com.

Dans un premier temps, nous nous proposons de faire une étude exploratoire pour se familiariser avec la donnée et comprendre comment les variables se comportent. Cette étape permettra egalement d'identifier d'eventuelles anomalies et de nettoyer la donnée si nécessaire.

Dans un second temps, nous nous attacherons à l'explicabilité des données, c'est à dire répondre à la question "Quelles sont les variables les plus importantes pour expliquer le prix d'un bien ?"

In [ ]:
# Librairie pour lire des fichiers de données tabulaires
import pandas as pd

In [ ]:
file_path = 'selogerdata.csv'
dataset = pd.read_csv(file_path)

In [ ]:
dataset.shape

(8899, 24)

Le jeu de données contient 8899 lignes et 24 colonnes.

In [ ]:
# Visualiser les 5 premières lignes du dataset
dataset.head()

number  codeinsee  codepostal     cp  etage  idagence  idannonce  idtiers  \
0      11     750116       75016  75016      0    178817  144013899   227883   
1       8     750110       75010  75010      0     12089  149750677     3765   
2       6     750118       75018  75018      0    115811  138648733   108873   
3      16     750110       75010  75010      0     12089  149750679     3765   
4       4     750116       75016  75016      0     97567  147153067   145098   

   idtypechauffage  idtypecommerce  ... nb_photos nb_pieces  position    prix  \
0                0               0  ...        13         3        11   59000   
1                0               0  ...         7         1         8   77500   
2                0               0  ...         8         2         6   94000   
3                0               0  ...         4         1        16   96000   
4                0               0  ...         5         2         4  100000   

   si_balcon  si_sdEau  si_sdbain  surface   typedebien         ville  
0          1         0          0    69.00  Appartement  Paris 16ème  
1          0         0          0     8.49  Appartement  Paris 10ème  
2          0         0          0    37.00  Appartement  Paris 18ème  
3          0         0          0    10.26  Appartement  Paris 10ème  
4          0         0          0    14.93  Appartement  Paris 16ème  

[5 rows x 24 columns]

## Nettoyage des données

On peut d'ores et déjà remarquer que les colonnes `idtypecyuisine` et `ville` doivent etre nettoyées

In [ ]:
# Afficher le nom des colonnes du dataset
dataset.columns

Index(['number', 'codeinsee', 'codepostal', 'cp', 'etage', 'idagence',
       'idannonce', 'idtiers', 'idtypechauffage', 'idtypecommerce',
       'idtypecuisine', 'idtypepublicationsourcecouplage', 'naturebien',
       'nb_chambres', 'nb_photos', 'nb_pieces', 'position', 'prix',
       'si_balcon', 'si_sdEau', 'si_sdbain', 'surface', 'typedebien', 'ville'],
      dtype='object')

In [ ]:
# Afficher la fréquence des types de cuisine.
dataset['idtypecuisine'].unique()

array(['équipée', 'aucune', '0', 'coin cuisine équipé',
       'coin cuisine', 'séparée', 'séparée équipée', 'américaine',
       'américaine équipée', '-1'], dtype=object)

Que veulent dire les valeurs 0 et -1 ?
On pourrait penser que 0 corresond à la situation ou il n'y a pas de cuisine mais ce n'est pas le cas en regardant les données. Nous allons créer des groupes à part pour ces catégories


In [ ]:
# Correction des données
mapper = {
    's√©par√©e': 'séparée',
    '0': 'categorie_0',
    's√©par√©e √©quip√©e': 'séparée équipée',
    'aucune': 'aucune',
    'am√©ricaine √©quip√©e': 'américaine équipée',
    '√©quip√©e': 'équipée',
    'am√©ricaine': 'américaine',
    'coin cuisine': 'coin cuisine',
    'coin cuisine √©quip√©': 'coin cuisine équipé',
    '-1': 'catégorie_1'
}

In [ ]:
# Appliquer les corrections
dataset['idtypecuisine'] = dataset['idtypecuisine'].map(mapper)

Nous allons également corriger les caractères accentués pour la variable ville.

In [ ]:
# Afficher les 10 premières villes
dataset['ville'].unique()

array(['Paris 16ème', 'Paris 10ème', 'Paris 18ème', 'Melun',
       'Corbeil Essonnes', 'Étampes', 'Issou', 'Limay', 'Évry', 'Avon',
       'Evry', 'Saint-Denis', 'Trappes', 'Ris Orangis', 'Paris 19ème',
       'Paris 14ème', 'Épinay-sous-Sénart', 'Eragny sur Oise',
       'Argenteuil', 'Bessancourt', 'Etampes', 'Montlhery', 'Gagny',
       'Fleury Merogis', 'Villeparisis', 'Fosses', 'Vaujours', 'Les Ulis',
       'Meaux', 'Pontoise', 'Cergy', 'Verneuil sur Seine',
       'Brie Comte Robert', 'Carrieres sous Poissy', 'Héricy', 'Igny',
       'Paris 20ème', 'Les Pavillons-sous-Bois', 'Mennecy',
       'Vert le Petit', 'Morsang-sur-Orge', 'Linas', 'Mandres les Roses',
       'Paris 17ème', 'Villepinte', 'Paris 2ème', 'Paris 15ème',
       'Paris 6ème', 'Paris 3ème', 'Paris 8ème', 'Paris 5ème',
       'Paris 11ème', 'Paris 9ème', 'Paris 13ème', 'Paris 12ème',
       'Garches', 'Chatou', 'Sucy en Brie', 'Louveciennes', 'Le Pecq',
       'Enghien les Bains', 'Saint Ger

In [ ]:
# Corriger les accents
dataset['ville'] =  (
    dataset['ville']
    .str.replace('√®', 'è')
    .str.replace('√â', 'É')
    .str.replace('√©', 'é')
    .str.replace('√¨', 'ê')
)

In [ ]:
dataset['ville'].unique()

array(['Paris 16ème', 'Paris 10ème', 'Paris 18ème', 'Melun',
       'Corbeil Essonnes', 'Étampes', 'Issou', 'Limay', 'Évry', 'Avon',
       'Evry', 'Saint-Denis', 'Trappes', 'Ris Orangis', 'Paris 19ème',
       'Paris 14ème', 'Épinay-sous-Sénart', 'Eragny sur Oise',
       'Argenteuil', 'Bessancourt', 'Etampes', 'Montlhery', 'Gagny',
       'Fleury Merogis', 'Villeparisis', 'Fosses', 'Vaujours', 'Les Ulis',
       'Meaux', 'Pontoise', 'Cergy', 'Verneuil sur Seine',
       'Brie Comte Robert', 'Carrieres sous Poissy', 'Héricy', 'Igny',
       'Paris 20ème', 'Les Pavillons-sous-Bois', 'Mennecy',
       'Vert le Petit', 'Morsang-sur-Orge', 'Linas', 'Mandres les Roses',
       'Paris 17ème', 'Villepinte', 'Paris 2ème', 'Paris 15ème',
       'Paris 6ème', 'Paris 3ème', 'Paris 8ème', 'Paris 5ème',
       'Paris 11ème', 'Paris 9ème', 'Paris 13ème', 'Paris 12ème',
       'Garches', 'Chatou', 'Sucy en Brie', 'Louveciennes', 'Le Pecq',
       'Enghien les Bains', 'Saint Germain en Laye', 'Paris 

## Analyse exploratoire des données
Nous allons commencer par un aperçu sur les données sous forme de résumé statitiques des différentes variables numériques.

In [ ]:
dataset.describe()

number      codeinsee    codepostal            cp   etage  \
count  8899.000000    8899.000000   8899.000000   8899.000000  8899.0   
mean      9.531296  751745.680189  75186.144735  75186.144735     0.0   
std       5.765614   16287.896954   1660.091189   1660.091189     0.0   
min       0.000000  750101.000000  75001.000000  75001.000000     0.0   
25%       5.000000  750108.000000  75008.000000  75008.000000     0.0   
50%      10.000000  750114.000000  75014.000000  75014.000000     0.0   
75%      15.000000  750116.000000  75017.000000  75017.000000     0.0   
max      19.000000  950500.000000  95880.000000  95880.000000     0.0   

            idagence     idannonce        idtiers  idtypechauffage  \
count    8899.000000  8.899000e+03    8899.000000           8899.0   
mean   124110.094280  1.473531e+08  136989.252500              0.0   
std    104338.907661  3.661055e+06  109732.960794              0.0   
min        33.000000  7.437284e+07      98.000000              0.0   
25%     39738.000000  1.463174e+08   26516.000000              0.0   
50%     93920.000000  1.486361e+08  123925.000000              0.0   
75%    203657.000000  1.495161e+08  213489.000000              0.0   
max    362103.000000  1.502498e+08  425123.000000              0.0   

       idtypecommerce   naturebien  nb_chambres    nb_photos    nb_pieces  \
count          8899.0  8899.000000  8899.000000  8899.000000  8899.000000   
mean              0.0     1.638723     2.319474     9.697269     3.838746   
std               0.0     0.931184     1.170877     4.511174     1.483164   
min               0.0     1.000000     1.000000     0.000000     1.000000   
25%               0.0     1.000000     1.000000     6.000000     3.000000   
50%               0.0     1.000000     2.000000     9.000000     4.000000   
75%               0.0     3.000000     3.000000    12.000000     5.000000   
max               0.0     3.000000    22.000000    27.000000    53.000000   

          position          prix    si_balcon  si_sdEau  si_sdbain  \
count  8899.000000  8.899000e+03  8899.000000    8899.0     8899.0   
mean      9.531296  1.210579e+06     0.271716       0.0        0.0   
std       5.765614  4.462923e+05     0.444869       0.0        0.0   
min       0.000000  5.900000e+04     0.000000       0.0        0.0   
25%       5.000000  7.950000e+05     0.000000       0.0        0.0   
50%      10.000000  1.475000e+06     0.000000       0.0        0.0   
75%      15.000000  1.490000e+06     1.000000       0.0        0.0   
max      19.000000  2.000000e+06     1.000000       0.0        0.0   

           surface  
count  8899.000000  
mean     96.070707  
std      38.127731  
min       0.000000  
25%      66.400000  
50%      93.000000  
75%     129.000000  
max     250.000000

Ce résumé statistique comporte déjà plusieurs enseignements :

* Les colonnes `etage`, `idtypechauffage`, `idtypecommerce`, `si_sdEau`,	`si_sdbain` ne comportent que des valeurs nulles. Elles n'apportent donc aucune information pour expliquer ou prédire le prix.

* Le jeu de données contient des outliers : on peut par exemple noter que le maximum du nombre de pièces est 53. De même le maximum du nombre de chambre est 22. Par ailleurs le minimum de la surface est 0.

Regardons ces outliers plus en détails.


In [ ]:
# sélectionner la ligne qui correspond au bien ayant 53 pièces


In [ ]:
# Sélectionner la ligne qui correspond au bien ayant 22 chambres


In [ ]:
# Sélectionner la ligne qui correspond au bien dont la surface est nulle


On va maitnenant enlever les valeurs aberrantes du dataset.

In [ ]:
# Supprimer les lignes aberrantes


Ragrdons maintenant la cardinalité des différentes variables. Ce qui nous intéresse ici, ce sont les variables catégorielles.

In [ ]:
# Combien de valeurs uniques possède chaque colonne ?
dataset.nunique().sort_values()

si_sdbain                             1
si_sdEau                              1
etage                                 1
idtypechauffage                       1
idtypecommerce                        1
si_balcon                             2
typedebien                            2
idtypepublicationsourcecouplage       3
naturebien                            3
nb_chambres                           6
nb_pieces                             8
idtypecuisine                        10
position                             20
number                               20
nb_photos                            27
cp                                  105
codepostal                          105
codeinsee                           107
ville                               112
prix                               1051
surface                            1217
idagence                           1412
idtiers                            1412
idannonce                          3812
dtype: int64

* Les variables `si_sdbain`, `si_sdEau`, `etage`, `idtypechauffage`, `idtypecommerce` ne comportent qu'une unique valeur, comme nous venons de le voir plus haut.
* `cp` et `codepostal` semblent se référer à la même variable. Par ailleurs, `codeinsee` et `ville` ont l'air de vehiculer plus ou moins la même information. En première approximation nous ne garderons qu'une seule de ces 4 variables.

* Les variables `position` et `number` ont le même nombre de modalités.

* Plus important peut-être, il y a 3815 modalités pour `idannonce` (annonces différentes) alors que le jeu de données contient 8899 lignes. On peut soupçonner ici la présence de doublons.

* Les variables `idagence` et`idtiers` ont le même nombre de modalités, très élevé. Il y a 1413 modalités pour chacune des variables alors qu'il n'y a potentiellement que 3815 annonces uniques, soit un peu plus de deux annonces par modalité de `idagence`.

Regardons ce dernier point plus en détails.



In [ ]:
idanonce_counts = dataset['idannonce'].value_counts()
display(idanonce_counts)
idanonce_counts.values[:100]

On peut noter la présence d'annonces qui se répètent 202 ou 53 fois. Il s'agit probablement de doublons dus à la fusions de plusieurs bases de données.

In [ ]:
# Supprimer les doublons
dataset.drop_duplicates(inplace = True)
dataset.shape

(3833, 24)

In [ ]:
# Vérifier si les colonnes "cp" et "codepostal" sont identiques
dataset['cp'].equals(dataset['codepostal'])

In [ ]:
# Vérifier si les colonnes "position" et "number" sont identiques
dataset['position'].equals(dataset['number'])

Nous allons maintnant faire le bilan des variables à supprimer.

* Les variables `si_sdbain`, `si_sdEau`, `etage`, `idtypechauffage`, `idtypecommerce` ne comportent que des valeurs nulles.
* Les variables `idtiers` et `idagence` renvoient vers la même information. Par ailleurs, elles comportent beaucoup de trop de modalités par rapport au nombre d'oberservations.
* Les variables `cp`, `codepostal` sont identiques et contiennent à peu près la même information que `codeinsee`. Cette information est déjà présente dans la viable `ville`.

In [ ]:
# Bilan des variables à supprimer
to_remove = [
    'si_sdbain', 'si_sdEau', 'etage',
    'idtypechauffage', 'idtypecommerce', 'codepostal',
    'number', 'idtiers', 'cp', 'codeinsee', 'idannonce', 'idagence']



In [ ]:
# Enlever les colonnes dans to_remove

dataset.drop(to_remove, axis = 1, inplace = True)
dataset.shape

(3833, 12)

Il est également nécessaire de distinguer entre les variables quantitatives et les variables catégorielles, qui seront analysées de manières différentes.

In [ ]:
categorical = [
    'si_balcon', 'typedebien', 'idtypepublicationsourcecouplage',
    'naturebien','idtypecuisine', 'ville'
]

numerical = [
    'nb_chambres', 'nb_pieces', 'nb_photos', 'surface', 'position'
]

## Analyse descriptive des variables quantitatives

Dans ce qui suit, nous allons analyser les distributions des variables ainsi que les corrélations entre les différentes variables explicatives abvec la variable `prix`. Nous allons utiliser ici les deux packages de visualisation en Python `matplotlib` et `seaborn`.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

Commençons d'abord par regarder la distribution de la variable d'intérêt `prix`.


In [ ]:
# densité de la variable prix sous forme de disribution

Regardons maintenant les distributions des différentes variables numériques.

Nous allons étudier maintenant les corrélation entre les variables numériques et la variable d'intérêt `prix`.

In [ ]:
# Tracer un corrélogramme des variables numériques


## Variables catégorielles

On peut commencer pas visualiser les variables qui ont un faible nombre de modalités en fonction de la variable `prix`.

Regardons maintenant les variables catégorielles ayant un nombre imortant de modalités. Commençons par la variable `ville`.

In [ ]:
# Afficher le nombre de modalités de la variable ville
dataset['ville'].nunique()

112

In [ ]:
# Répartition des observations sur les modalités
dataset['ville'].value_counts()

ville
Paris 16ème                443
Paris 18ème                308
Paris 15ème                299
Paris 17ème                294
Paris 13ème                278
                          ... 
Mennecy                      1
Les Pavillons-sous-Bois      1
Igny                         1
Héricy                       1
Clichy                       1
Name: count, Length: 112, dtype: int64

In [ ]:
# Afficher la répartition des observations pour les modalités qui ne commencent pas par "Paris"
dataset[~dataset['ville'].str.startswith('Paris')]['ville'].value_counts()

ville
Corbeil Essonnes           5
Saint Germain en Laye      3
Le Pecq                    3
Villepinte                 3
Suresnes                   3
                          ..
Vert le Petit              1
Mennecy                    1
Les Pavillons-sous-Bois    1
Igny                       1
Clichy                     1
Name: count, Length: 92, dtype: int64

Nous voyons ici que les villes situées en dehors de Paris sont très faiblement représentées dans le jeu de données. Nous allons les regrouper dans une seule modalité "banlieue".

In [ ]:
# Regrouper les observations dans une variable "banlieue"
dataset['ville'] = dataset['ville'].apply(lambda x: x if x.startswith('Paris') else 'Banlieue')
dataset['ville'].value_counts()

ville
Paris 16ème    443
Paris 18ème    308
Paris 15ème    299
Paris 17ème    294
Paris 13ème    278
Paris 19ème    212
Paris 11ème    195
Paris 14ème    194
Paris 20ème    194
Paris 10ème    167
Paris 6ème     164
Paris 12ème    162
Paris 7ème     144
Paris 8ème     129
Banlieue       120
Paris 5ème     115
Paris 4ème     110
Paris 9ème     100
Paris 3ème      96
Paris 2ème      65
Paris 1er       44
Name: count, dtype: int64

In [ ]:
# Visualiser le prix en fonction des modalités de la variable ville

In [ ]:
# Visualiser le prix en fonction de la variable idtypecuisine